ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -1030,6 +1037,315 @@ private Pair<ServerErrorCode, GetResponse> getBlobFromRemoteReplica(ReplicateBlo
+        || replicateBlobRequest.getOperationType() == RequestOrResponseType.DeleteRequest) {
+      // For RequestOrResponseType.PutRequest, we don't care about the life version.
+      if (replicateBlobRequest.getLifeVersion() != MessageInfo.LIFE_VERSION_FROM_FRONTEND) {
+        logger.info("ReplicateBlobRequest invalid life version {}", replicateBlobRequest);
+        completeReplicateRequest(request, replicateBlobRequest, ServerErrorCode.Bad_Request, startProcessTime);
+        return;
+      }
+    }
+
+    // this replica itself is the source replica. Don't need do anything. Return success status.
+    DataNodeId remoteDataNode = clusterMap.getDataNodeId(remoteHostName, remoteHostPort);
+    if (remoteDataNode != null && remoteDataNode.equals(currentNode)) {
+      logger.info("ReplicateBlobRequest this is the source replica, return immediately. {}", replicateBlobRequest);
+      completeReplicateRequest(request, replicateBlobRequest, ServerErrorCode.No_Error, startProcessTime);
+      return;
+    }
+
+    // check if local store has the key already
+    ServerErrorCode errorCode = ServerErrorCode.No_Error;
+    MessageInfo localMessageInfo = null;
+    try {
+      StoreKey convertedKey = getConvertedStoreKeys(Collections.singletonList(blobId)).get(0);
+      if (!((BlobId) convertedKey).equals(blobId)) {
+        logger.info("ReplicateBlobRequest replicated Blob {} convert to local key {}", blobId, (BlobId) convertedKey);
+        blobId = (BlobId) convertedKey;
+      }
+      Store store = storeManager.getStore(blobId.getPartition());
+      localMessageInfo = store.findKey(convertedKey);
+      logger.info("ReplicateBlobRequest replicated Blob {}, local Store has the Key", blobId);
+    } catch (StoreException e) {
+      // store.findKey throws e.getErrorCode() == StoreErrorCodes.ID_Not_Found if it doesn't exist.
+      if (e.getErrorCode() == StoreErrorCodes.ID_Not_Found) {
+        errorCode = ServerErrorCode.No_Error;
+      } else {
+        logger.error("ReplicateBlobRequest store.findKey throw exception {} on Blob {},", e.getErrorCode(), blobId);
+        errorCode = ServerErrorCode.Unknown_Error;
+      }
+    } catch (Exception e) {
+      // getConvertedStoreKeys which may throw Exception
+      logger.error("ReplicateBlobRequest getConvertedStoreKeys unknown exception to replicate {} of {}", blobId,
+          replicateBlobRequest, e);
+      errorCode = ServerErrorCode.Unknown_Error;
+    }
+    if (errorCode != ServerErrorCode.No_Error || (
+        replicateBlobRequest.getOperationType() == RequestOrResponseType.PutRequest && localMessageInfo != null)) {
+      // If we want to replicate PutBlob and local host has the key already, return.
+      completeReplicateRequest(request, replicateBlobRequest, errorCode, startProcessTime);
+      return;
+    }
+
+    // Step 1: Repair the PutBlob if local store doesn't have the blob when repair type is RequestOrResponseType.PutRequest or RequestOrResponseType.TtlUpdateRequest
```

Comment:
QQ: the step 1, 2, 3 will not happen for same replicateBlobRequest right? 
if so, looks like step1,2,3 is more like case 1,2,3, so can we do something like:
if (case1)
else if (case2)
else if(case3)
So we can have just one "completeReplicateRequest(request, replicateBlobRequest, errorCode, startProcessTime)" at the end?

Reply:
One repair type may run multiple steps.
For example, for TTLUpdate, in order to repair it and if we don't have the PutBlob in the local store, we need repair PutBlob first. After that, we'll repair TtlUpdate.

Reply:
It seems like both putBlob repair and ttlUpdate repair logic are in repairPutBlob method. So I guess for one replicateBlobRequest, it will only have one type of OperationType and it won't be changed through handleReplicateBlobRequestV2 right? in that case, can we use if, else if logic so we can put  completeReplicateRequest at the bottom only?

Reply:
Got it. thx for the explanation!